retrievalmarket/impl/client_test.go

```diff
@@ -105,10 +109,78 @@ func TestClient_Query(t *testing.T) {
 		net := tut.NewTestRetrievalMarketNetwork(tut.TestNetworkParams{
 			QueryStreamBuilder: qsbuilder,
 		})
-		c := retrievalimpl.NewClient(net, bs, testnodes.NewTestRetrievalClientNode(testnodes.TestRetrievalClientNodeParams{}))
+		c := retrievalimpl.NewClient(
+			net,
+			bs,
+			testnodes.NewTestRetrievalClientNode(testnodes.TestRetrievalClientNodeParams{}),
+			&testPeerResolver{})
 
 		statusCode, err := c.Query(ctx, rpeer, pcid, retrievalmarket.QueryParams{})
 		assert.EqualError(t, err, "query response failed")
 		assert.Equal(t, retrievalmarket.QueryResponseUndefined, statusCode)
 	})
 }
+
+func TestClient_FindProviders(t *testing.T) {
+	bs := bstore.NewBlockstore(dss.MutexWrap(datastore.NewMapDatastore()))
+	expectedPeer := peer.ID("somevalue")
+
+	var qsb tut.QueryStreamBuilder = func(p peer.ID) (rmnet.RetrievalQueryStream, error) {
+		return tut.NewTestRetrievalQueryStream(tut.TestQueryStreamParams{
+			Writer:     tut.TrivialQueryWriter,
+			RespReader: tut.TrivialQueryResponseReader,
+		}), nil
+	}
+	net := tut.NewTestRetrievalMarketNetwork(tut.TestNetworkParams{
+		QueryStreamBuilder: tut.ExpectPeerOnQueryStreamBuilder(t, expectedPeer, qsb, "Peers should match"),
+	})
+
+	t.Run("when providers are found, returns providers", func(t *testing.T) {
+		peers := tut.RequireGenerateRetrievalPeers(t, 3)
+		testResolver := testPeerResolver{peers: peers}
+
+		c := retrievalimpl.NewClient(net, bs, &testRetrievalNode{}, &testResolver)
+		testCid := []byte("somePieceCID")
+		assert.Len(t, c.FindProviders(testCid), 3)
+	})
+
+	t.Run("when there is an error, returns empty provider list", func(t *testing.T) {
+		testResolver := testPeerResolver{peers: []retrievalmarket.RetrievalPeer{}, resolverError: errors.New("boom")}
+		c := retrievalimpl.NewClient(net, bs, &testRetrievalNode{}, &testResolver)
+		badCid := []byte("doesn't matter")
+		assert.Len(t, c.FindProviders(badCid), 0)
+	})
+
+	t.Run("when there are no providers", func(t *testing.T) {
+		testResolver := testPeerResolver{peers: []retrievalmarket.RetrievalPeer{}}
+		c := retrievalimpl.NewClient(net, bs, &testRetrievalNode{}, &testResolver)
+		testCid := []byte("unimportant")
+		assert.Len(t, c.FindProviders(testCid), 0)
+	})
+}
+
+type testRetrievalNode struct {
```

Comment:
sorry hate to be a stickeler here, but unless there is a compelling reason, this should use the existing TestRetrievalClientNode -- there's no need to make a dupe here.

Reply:
I had this in my branch before you merged yours to master, but agree there should not be two of them.  I overlooked that you had put one in.

Reply:
I just missed that it had been moved out of this file when I rebased, is all.